application/controllers/QuestionAdministrationController.php

```diff
@@ -2278,31 +2278,44 @@ function ($oAnswerOption) {
      *
      * @param int $iQuestionId
      * @param string $sQuestionType
-     * @param string $question_template
+     * @param string $sQuestionTheme
      * @return array
      * @throws CException
      * @throws Exception
      */
-    private function getAdvancedOptions($iQuestionId = null, $sQuestionType = null, $question_template = 'core')
+    private function getAdvancedOptions($iQuestionId = null, $sQuestionType = null, $sQuestionTheme = 'core')
     {
         //here we get a Question object (also if question is new --> QuestionCreate)
-
         $oQuestion = $this->getQuestionObject($iQuestionId, $sQuestionType);
-        $advancedSettings = $oQuestion->getAdvancedSettingsWithValuesByCategory(null);
-        // TODO: Why can empty array be saved as value?
-        foreach ($advancedSettings as &$category) {
-            foreach ($category as &$setting) {
-                if ($setting['value'] === []) {
-                    $setting['value'] = null;
-                }
-            }
-        }
+
+        // Get the advanced settings array
+        $advancedSettings = $oQuestion->getAdvancedSettingsWithValues(null, $sQuestionTheme);
+
+        // Group the array in categories
+        $advancedSettings = self::groupAttributesByCategory($advancedSettings);
+
         // This category is "general setting".
         unset($advancedSettings['Attribute']);
 
         return $advancedSettings;
     }
 
+    /**
+     * Receives an array of question attributes and groups them by category.
+     * Used by advanced settings widget.
+     *
+     * @param array $aAttributes
+     * @return array Grouped question attributes, with category as array key
+     */
+    private static function groupAttributesByCategory($aAttributes)
```

Comment:
Doesn't need to be static? Static functions can't be mocked and should be avoided in case they're really needed (like for factory methods).

Reply:
Honestly, it was non-static initially. But it felt very wrong, as this function don't use anything from the object. Is a helper method.
Anyway, will change it

Reply:
Yeah, if you have no "$this", another alternative is to move it to a normal function (if it has no side-effects, at least; for a discussion of what is considered a side-effect, see: https://manual.limesurvey.org/Code_quality_guide#Push_side-effects_up_in_the_stack_trace).

Reply:
> One thing: Please never add static methods in ActiveRecord classes. In the end, we want ActiveRecord to _only_ contain the database relation. Can you move them to a service class, please? Thank you. Reasons are as before:
> 
> 1. One responsibility per class (database interaction is one)
> 2. Testability (static methods can't easily be mocked)
> 3. Readability (big classes with lots of different behaviour are harder to get a conception of)

Yes, we were focused on unwinding the code/
Probably the best is to create a service class.